File: apps/api/src/ensemble/inference.ts
import type { Action, Probs } from '../domain/signal.js';

/** Convierte el score neto en probabilidades BUY/HOLD/SELL vía softmax con temperatura. */
export interface MacroModulation {
  bias: number;
  wMacro: number;
}

export function inferProbs(
  net: number,
  temperature: number,
  holdBand: number,
  macro?: MacroModulation,
): Probs {
  const t = temperature > 0 ? temperature : 0.5;
  const macroTerm = macro ? macro.wMacro * macro.bias : 0;
  const logits = { BUY: net / t + macroTerm, SELL: -net / t - macroTerm, HOLD: holdBand / t };
  const max = Math.max(logits.BUY, logits.SELL, logits.HOLD);
  const e = {
    BUY: Math.exp(logits.BUY - max),
    SELL: Math.exp(logits.SELL - max),
    HOLD: Math.exp(logits.HOLD - max),
  };
  const sum = e.BUY + e.SELL + e.HOLD;
  return { BUY: e.BUY / sum, HOLD: e.HOLD / sum, SELL: e.SELL / sum };
}

export function pickAction(probs: Probs): { action: Action; confidence: number } {
  const entries: Array<[Action, number]> = [
    ['BUY', probs.BUY],
    ['HOLD', probs.HOLD],
    ['SELL', probs.SELL],
  ];
  entries.sort((a, b) => b[1] - a[1]);
  const top = entries[0] ?? (['HOLD', 1] as [Action, number]);
  return { action: top[0], confidence: top[1] };
}

export function confluence(net: number, bias: number): 'aligned' | 'conflict' | 'neutral' {
  if (net === 0 || bias === 0) return 'neutral';
  return Math.sign(net) === Math.sign(bias) ? 'aligned' : 'conflict';
}

/**
 * Escalado de w_macro por temporalidad (M1, estructura preparada · DESACTIVADA por defecto).
 * Cuando `enableScaling` es false devuelve el w_macro sin cambios. Se activará al reintroducir
 * el análisis fundamental, ponderando el sesgo macro menos en temporalidades cortas.
 */
export function scaledWMacro(
  wMacro: number,
  interval: string,
  cfg: { enableScaling: boolean; tfScale: Record<string, number> },
): number {
  if (!cfg.enableScaling) return wMacro;
  const factor = cfg.tfScale[interval] ?? 1;
  return wMacro * factor;
}

File: apps/quant/trademe_quant/inference.py
"""Mirror de la inferencia softmax con modulación macro (paridad con apps/api)."""

from __future__ import annotations

import math


def infer_probs(
    net: float,
    temperature: float,
    hold_band: float,
    macro_bias: float | None = None,
    w_macro: float = 0.0,
    independence: float = 1.0,
    fund_term: float = 0.0,
) -> dict[str, float]:
    """Softmax con modulación macro, penalización fundamental y desinflado por dependencia.

    `independence` escala los TRES logits por igual (ver trademe_quant.independence). Como escalar
    todos los logits por una constante positiva no altera cuál es el mayor, el ajuste **no cambia la
    dirección de la decisión**: solo aplana la distribución y baja la confianza declarada. Es una
    corrección de calibración, no de criterio.

    `fund_term` (M12) es lo contrario: **sí cambia el criterio, y solo en un lado**. Se resta al
    logit BUY y no toca el de SELL, porque el efecto medido del funding existe únicamente en los
    largos (ver trademe_quant.fundamental). Vale 0 mientras el score esté en sombra.
    """
    t = temperature if temperature > 0 else 0.5
    macro_term = w_macro * macro_bias if macro_bias is not None else 0.0
    k = independence if independence > 0 else 1.0
    logits = {
        "BUY": k * (net / t + macro_term - fund_term),
        "SELL": k * (-net / t - macro_term),  # asimetría: el funding no informa los cortos
        "HOLD": k * (hold_band / t),
    }
    peak = max(logits.values())
    exp = {k: math.exp(v - peak) for k, v in logits.items()}
    total = sum(exp.values())
    return {k: v / total for k, v in exp.items()}


def pick_action(probs: dict[str, float]) -> str:
    return max(probs, key=lambda k: probs[k])


def scaled_w_macro(w_macro: float, interval: str, cfg: dict[str, object]) -> float:
    """Escalado de w_macro por temporalidad (M1, estructura preparada · DESACTIVADA por defecto).

    Si enable_scaling es False devuelve w_macro sin cambios. Mirror de apps/api inference.ts.
    """
    if not cfg.get("enable_scaling", False):
        return w_macro
    tf_scale = cfg.get("tf_scale", {})
    factor = float(tf_scale.get(interval, 1.0)) if isinstance(tf_scale, dict) else 1.0
    return w_macro * factor
